### PR TITLE
[wip] gradient accumulation, another attempt

### PR DESCRIPTION
**Pull request by karpathy**
Doesn't work. On master, we reproduce our Python script (almost) exactly by running:

```
make train_gpt2cu NO_MULTI_GPU=1 USE_CUDNN=1
./train_gpt2cu -b 4 -t 64 -l 1e-4 -v 200 -s 200 -a 1 -x 10
```

But doing this on this branch fails and the loss stagnates, something is wrong with the optimization.
Running an actual training loop, the loss does not decrease.

debugging...

### CODE REVIEW

**Comment by karpathy** [comment]
(The Python version works and behaves as expected. E.g. bumping the total batch size from 256 to 512 or 1024 produces exactly identical results, as it should be, because the batch is fixed so doing gradient accumulation over many fwd bwd passes over it create all the same gradients, so the mean of those doesn't change and everything remains the same)

**Comment by karpathy** [comment]
Issue was with streams, which were complex and aren't adding much perf anyway atm.
Threw away all streams, back to default 1 stream for now. 
Will be more careful re-adding streams in future commits.
Gradient accumulation seems to work ok now.
I did some debugging a bit with commands such as:

```bash
# fp32 test
# make train_gpt2cu NO_MULTI_GPU=1 PRECISION=FP32
# ./train_gpt2cu -b 4 -t 64 -l 1e-4 -v 200 -s 200 -a 1 -x 10 -f 0 -e gpt2_124M.bin

# bf16 single gpu test
# make train_gpt2cu NO_MULTI_GPU=1 USE_CUDNN=1
# ./train_gpt2cu -b 4 -t 64 -l 1e-4 -v 200 -s 200 -a 1 -x 10

# make train_gpt2cu USE_CUDNN=1
make train_gpt2cu USE_CUDNN=1
mpirun -np 4 ./train_gpt2cu -b 4 -t 64 -l 1e-4 -v 200 -s 200 -a 1 -x 10 -d 4096
```

And the way I tested things was to use the new `-d` flag to force gradient accumulation. e.g. if `-d 512`, then the total batch size is 512, forcing gradient accumulation to be 2. And because the batches here are hacked to be all same, you expect the same results exactly, because all the accumulations are producing the same gradients, and then averaging out to the identical point. (Note that for the multi-gpu test I had to also hack this `loader->process_rank = 0` inside dataloader init, otherwise the dataloader tries to be different on each process).
